_posts/2015-12-14-simply-secure-rails-api.markdown

```diff
@@ -0,0 +1,499 @@
+
+```ruby
+module Api
+  module V1
+    class LoginController < ApiController
+      
+      def create # POST: /api/v1/login
+        @user = User.find_by(email: params[:email])
+
+        if @user.present? && @user.valid_password?(params[:password])
+          respond_with: @user.as_json()
+        else 
+          head 404
+        end
+      end
+
+    end
+  end
+end
+```
+
+##### It has a users controller
+
+```ruby
+module Api
+  module V1
+    class UsersController < ApiController
+      
+      def show # GET: /api/v1/users/[UUID]
+        @user = User.find_by(uuid: params[:id])
+
+        if @user.present?
+          respond_with: @user.as_json()
+        else 
+          head 404
+        end
+      end
+
+    end
+  end
+end
+```
+
+---
+
+##### Application 2: 
+Social Network (https://network.wework.io)
+A rails web application which uses the aforementioned authentication service to log a user in and seed user data.
+
+We want this application to log a user in, persist a "copy" of this user (like any standard OAuth app) and allow that user to create posts.
+
+#### Let's get started
+
+First you want to add a method to the authentication service's User Model. This method encrypts a
+user's uuid.
+
+Using ruby's OpenSSL library
+
+```ruby
+class User < ActiveRecord::Base
+  # Schema
+  # Name: string
+  # UUID: UUID (primary key)
+  # Email
+  # Devise password stuff
+  devise :database_authenticatable
+
+  def encrypted
+    cipher         =  OpenSSL::Cipher.new('aes-256-cbc')
+    cipher.encrypt
+    cipher.key     =  ENV['SOME_SECRET_KEY']
+    cipher.iv      =  ENV['SOME_SECRET_IV']
+    cipher.padding =  1
+    encrypted      =  cipher.update(self.uuid) + cipher.final # use only this user's UUID for encryption
+    Base64.urlsafe_encode64(encrypted).encode('utf-8')
+  end
+end
+```
+
+Look at this bonkers method! It's pretty hard to figure out what's going on here. And to be fair it's outside 
+the norm of ordinary, MVC style web development. So lets go through it line by line.
+
+
+#### Line 1
+```ruby
+cipher =  OpenSSL::Cipher.new('aes-256-cbc')
+```
+
+The argument 'aes-256-cbc' is a hyphenated description of components of the cipher. 
+
+1. "aes" is the name of the encryption algorithm. In this case, its an acronym for [Advanced Encryption Standard](http://aesencryption.net/).
+
+2. 256 is the number of bits of the key used in the algorithm. You'll usually see 128 or 256. Practically, this means that the key is "large" enough that it cannot be feasibly broken via a brute force attack (it would take foooooorrreeeevvvveeeerrr) . 
+
+3. cbc stands for "cipher block chaining" and is the mode by which the aes algorithm encrypts the data. Its most commonly compared to the ecb [electronic code book](http://searchsecurity.techtarget.com/definition/Electronic-Code-Book). 
+Its generally preferred to use cbc because ebc may expose pieces of the actual key in the encrypted string when encrypting smaller pieces of data. 
```

Comment:
The main reason cbc is preferred is because with ecb identical blocks in the text will have identical output, this allowing information to be extracted.